src/asio_service.cxx

```diff
@@ -1124,13 +1151,38 @@ class asio_rpc_client
             when_done(rsp, except);
             return;
         }
-        send(req, when_done, send_timeout_ms);
+        register_req_send(req, when_done, send_timeout_ms);
     }
 
     virtual void send(ptr<req_msg>& req,
                       rpc_handler& when_done,
                       uint64_t send_timeout_ms = 0) __override__
     {
+        if (impl_->get_options().streaming_mode_) {
+            pre_send(req, when_done, send_timeout_ms);
+        } else {
+            register_req_send(req, when_done, send_timeout_ms);
+        }
+    }
+
+    void pre_send(ptr<req_msg>& req,
+                    rpc_handler& when_done,
+                    uint64_t send_timeout_ms) {
+        auto_lock(pending_write_reqs_lock_);
+        if (pending_write == 0) {
```

Comment:
When the queue is empty, the coming request will be sent immediately and not be added to this queue. So when the queue is empty there are two cases. One is no flying request and another is we have one flying request.

Reply:
In close_socket, the when_done of the request that triggered the error have already been invoked. So to avoid call the when_done twice. Request that being processed is not stored in the queue.

Reply:
I understand your point about this variable, but maintaining a separate counter (which has different value from `pending_write_reqs_.size()`) will be very confusing and mistake-prone during the future maintenance.

Let's do this way; put the current in-flight request into the queue as well:
```
    void pre_send(ptr<req_msg>& req,
                  rpc_handler& when_done,
                  uint64_t send_timeout_ms) {
        auto_lock(pending_write_reqs_lock_);
        pending_write_reqs_.push_back(
            cs_new<pending_req_pkg>(req, when_done, send_timeout_ms));
        if (pending_write_reqs_.size() == 1) {
            register_req_send(req, when_done, send_timeout_ms);
        }
        ...
    }
```
and
```
    void post_send(ptr<req_msg>& req, rpc_handler& when_done) {
	    ...
        // next process write
        {
            auto_lock(pending_write_reqs_lock_);
            pending_write_reqs_.pop_front();
            if (pending_write_reqs_.size() > 0) {
                ptr<pending_req_pkg> next_req_pkg = *pending_write_reqs_.begin();
                register_req_send(next_req_pkg->req_, next_req_pkg->when_done_, 
                                  next_req_pkg->timeout_ms_);                
                p_db("trigger next write, start_log_idx: %ld, pending write reqs: %ld",
                     next_req_pkg->req_->get_last_log_idx(), pending_write_);
            }
        }
    }
```
same to read requests as well. This also resolves the issue with `when_done()` and `close_socket()`; when error happens, if `streaming_mode_` is on, we just call `close_socket()` without calling `when_done()`.